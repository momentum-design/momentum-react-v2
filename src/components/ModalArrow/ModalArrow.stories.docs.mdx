The `<ModalArrow />` component. This component is designed to be strictly consumbed by the
`<Modal />` group of components.
