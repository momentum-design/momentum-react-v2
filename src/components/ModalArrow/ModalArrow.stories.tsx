import { MultiTemplate, Template } from '../../storybook/helper.stories.templates';
import { DocumentationPage } from '../../storybook/helper.stories.docs';
import StyleDocs from '../../storybook/docs.stories.style.mdx';

import ModalArrow, { ModalArrowProps, MODAL_ARROW_CONSTANTS as CONSTANTS } from './';
import argTypes from './ModalArrow.stories.args';
import Documentation from './ModalArrow.stories.docs.mdx';

export default {
  title: 'Momentum UI/ModalArrow',
  component: ModalArrow,
  parameters: {
    expanded: true,
    docs: {
      page: DocumentationPage(Documentation, StyleDocs),
    },
  },
};

const Example = Template<ModalArrowProps>(ModalArrow);

Example.argTypes = { ...argTypes };

Example.args = {
  color: CONSTANTS.COLORS.TERTIARY,
  placement: CONSTANTS.PLACEMENTS.BOTTOM,
};

const Colors = MultiTemplate<ModalArrowProps>(ModalArrow);

Colors.argTypes = { ...argTypes };
delete Colors.argTypes.color;

Colors.args = {
  placement: CONSTANTS.PLACEMENTS.BOTTOM,
};

Colors.parameters = {
  variants: [
    ...Object.values(CONSTANTS.COLORS).map((color) => ({ color, style: { margin: '1rem' } })),
  ],
};

const Placements = MultiTemplate<ModalArrowProps>(ModalArrow);

Placements.argTypes = { ...argTypes };
delete Placements.argTypes.placement;

Placements.args = {
  color: CONSTANTS.COLORS.TERTIARY,
};

Placements.parameters = {
  variants: [
    ...Object.values(CONSTANTS.PLACEMENTS).map((placement) => ({
      placement,
      style: { margin: '1rem' },
    })),
  ],
};

const Common = MultiTemplate<ModalArrowProps>(ModalArrow);

Common.argTypes = { ...argTypes };
delete Common.argTypes.color;
delete Common.argTypes.placement;

Common.parameters = {
  variants: [
    ...Object.values(CONSTANTS.PLACEMENTS).map((placement) =>
      Object.values(CONSTANTS.COLORS).map((color) => ({
        color,
        placement,
        style: { margin: '1rem' },
      }))
    ),
  ].flat(),
};

export { Example, Colors, Placements, Common };
