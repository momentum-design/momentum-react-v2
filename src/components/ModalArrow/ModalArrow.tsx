import React, { FC } from 'react';
import classnames from 'classnames';

import { DEFAULTS, PLACEMENTS, STYLE } from './ModalArrow.constants';
import type { Props } from './ModalArrow.types';
import './ModalArrow.style.scss';

/**
 * The ModalArrow component. This is designed to be strictly consumed by the ModalContainer component.
 */
const ModalArrow: FC<Props> = (props: Props) => {
  const { className, color, id, placement = DEFAULTS.PLACEMENT, style } = props;

  // Point Reference.
  const pr = {
    end: '',
    start: '',
    tipEnd: '',
    tipPeak: '',
    tipStart: '',
  };

  const isVertical =
    placement.startsWith(PLACEMENTS.TOP) || placement.startsWith(PLACEMENTS.BOTTOM);

  switch (placement) {
    case PLACEMENTS.BOTTOM:
    case PLACEMENTS.BOTTOM_START:
    case PLACEMENTS.BOTTOM_END:
      pr.start = '0 12';
      pr.tipStart = '10 -10';
      pr.tipPeak = '2 -2';
      pr.tipEnd = '4 0';
      pr.end = '10 10';
      break;

    case PLACEMENTS.LEFT:
    case PLACEMENTS.LEFT_START:
    case PLACEMENTS.LEFT_END:
      pr.start = '0 0';
      pr.tipStart = '10 10';
      pr.tipPeak = '2 2';
      pr.tipEnd = '0 4';
      pr.end = '-10 10';
      break;

    case PLACEMENTS.RIGHT:
    case PLACEMENTS.RIGHT_START:
    case PLACEMENTS.RIGHT_END:
      pr.start = '12 24';
      pr.tipStart = '-10 -10';
      pr.tipPeak = '-2 -2';
      pr.tipEnd = '0 -4';
      pr.end = '10 -10';
      break;

    case PLACEMENTS.TOP:
    case PLACEMENTS.TOP_START:
    case PLACEMENTS.TOP_END:
      pr.start = '24 0';
      pr.tipStart = '-10 10';
      pr.tipPeak = '-2 2';
      pr.tipEnd = '-4 0';
      pr.end = '-10 -10';
      break;
  }

  const pathData = `m ${pr.start} l ${pr.tipStart} q ${pr.tipPeak} ${pr.tipEnd} l ${pr.end}`;

  const viewBoxHeight = isVertical ? 12 : 24;
  const viewBoxWidth = isVertical ? 24 : 12;
  const viewBox = `0 0 ${viewBoxWidth} ${viewBoxHeight}`;

  return (
    <svg
      id={id}
      className={classnames(STYLE.svg, className)}
      style={style}
      aria-hidden='true'
      xmlns="http://www.w3.org/svg/2000"
      width={viewBoxWidth}
      height={viewBoxHeight}
      viewBox={viewBox}
      data-placement={placement}
    >
      <defs>
        <clipPath id={`modal-arrow-cut-stroke-${placement}`}>
          <path d={pathData} />
        </clipPath>
      </defs>

      <path
        data-color={color || DEFAULTS.COLOR}
        clipPath={`url(#modal-arrow-cut-stroke-${placement})`}
        d={pathData}
      />
    </svg>
  );
};

export default ModalArrow;
