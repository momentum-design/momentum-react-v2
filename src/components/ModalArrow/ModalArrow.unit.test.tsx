import React from 'react';
import { mount } from 'enzyme';

import ModalArrow, { MODAL_ARROW_CONSTANTS as CONSTANTS } from './';

describe('<ModalArrow />', () => {
  const placement = CONSTANTS.PLACEMENTS.BOTTOM;

  describe('snapshot', () => {
    it('should match snapshot', () => {
      expect.assertions(1);

      const container = mount(<ModalArrow placement={placement} />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with className', () => {
      expect.assertions(1);

      const className = 'example-class';

      const container = mount(<ModalArrow placement={placement} className={className} />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with aria-hidden', () => {
      expect.assertions(1);

      const ariaHidden = 'true';

      const container = mount(<ModalArrow placement={placement} aria-hidden={ariaHidden} />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with id', () => {
      expect.assertions(1);

      const id = 'example-id';

      const container = mount(<ModalArrow placement={placement} id={id} />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with style', () => {
      expect.assertions(1);

      const style = { color: 'pink' };

      const container = mount(<ModalArrow placement={placement} style={style} />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with color', () => {
      expect.assertions(1);

      const color = Object.values(CONSTANTS.COLORS).pop();

      const container = mount(<ModalArrow color={color} placement={placement} />);

      expect(container).toMatchSnapshot();
    });
  });

  describe('attributes', () => {
    it('should have its svg class', () => {
      expect.assertions(1);

      const element = mount(<ModalArrow placement={placement} />)
        .find(ModalArrow)
        .getDOMNode();

      expect(element.classList.contains(CONSTANTS.STYLE.svg)).toBe(true);
    });

    it('should have ariaHidden prop in svg', () => {
      expect.assertions(1);

      const ariaHidden = 'true';

      const element = mount(<ModalArrow placement={placement} />)
        .find(ModalArrow)
        .getDOMNode();

      expect(element.getAttribute('aria-hidden')).toBe(ariaHidden);
    });

    it('should have provided class when className is provided', () => {
      expect.assertions(1);

      const className = 'example-class';

      const element = mount(<ModalArrow className={className} placement={placement} />)
        .find(ModalArrow)
        .getDOMNode();

      expect(element.classList.contains(className)).toBe(true);
    });

    it('should have provided id when id is provided', () => {
      expect.assertions(1);

      const id = 'example-id';

      const element = mount(<ModalArrow id={id} placement={placement} />)
        .find(ModalArrow)
        .getDOMNode();

      expect(element.id).toBe(id);
    });

    it('should have provided style when style is provided', () => {
      expect.assertions(1);

      const style = { color: 'pink' };
      const styleString = 'color: pink;';

      const element = mount(<ModalArrow style={style} placement={placement} />)
        .find(ModalArrow)
        .getDOMNode();

      expect(element.getAttribute('style')).toBe(styleString);
    });

    it('should have provided data-placement on path when placement is provided', () => {
      expect.assertions(1);

      const placement = Object.values(CONSTANTS.PLACEMENTS).pop();

      const element = mount(<ModalArrow placement={placement} />)
        .find(ModalArrow)
        .getDOMNode();

      expect(element.getAttribute('data-placement')).toBe(placement);
    });

    it('should have provided data-color when color is provided', () => {
      expect.assertions(1);

      const color = Object.values(CONSTANTS.COLORS).pop();

      const element = mount(<ModalArrow color={color} placement={placement} />)
        .find(ModalArrow)
        .getDOMNode()
        .getElementsByTagName('path')[1];

      expect(element.getAttribute('data-color')).toBe(color);
    });
  });
});
