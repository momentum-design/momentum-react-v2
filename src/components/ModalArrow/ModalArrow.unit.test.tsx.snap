// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ModalArrow /> snapshot should match snapshot 1`] = `
<ModalArrow
  placement="bottom"
>
  <svg
    aria-hidden="true"
    className="md-modal-arrow-svg"
    data-placement="bottom"
    height={12}
    viewBox="0 0 24 12"
    width={24}
    xmlns="http://www.w3.org/svg/2000"
  >
    <defs>
      <clipPath
        id="modal-arrow-cut-stroke-bottom"
      >
        <path
          d="m 0 12 l 10 -10 q 2 -2 4 0 l 10 10"
        />
      </clipPath>
    </defs>
    <path
      clipPath="url(#modal-arrow-cut-stroke-bottom)"
      d="m 0 12 l 10 -10 q 2 -2 4 0 l 10 10"
      data-color="primary"
    />
  </svg>
</ModalArrow>
`;

exports[`<ModalArrow /> snapshot should match snapshot with aria-hidden 1`] = `
<ModalArrow
  aria-hidden="true"
  placement="bottom"
>
  <svg
    aria-hidden="true"
    className="md-modal-arrow-svg"
    data-placement="bottom"
    height={12}
    viewBox="0 0 24 12"
    width={24}
    xmlns="http://www.w3.org/svg/2000"
  >
    <defs>
      <clipPath
        id="modal-arrow-cut-stroke-bottom"
      >
        <path
          d="m 0 12 l 10 -10 q 2 -2 4 0 l 10 10"
        />
      </clipPath>
    </defs>
    <path
      clipPath="url(#modal-arrow-cut-stroke-bottom)"
      d="m 0 12 l 10 -10 q 2 -2 4 0 l 10 10"
      data-color="primary"
    />
  </svg>
</ModalArrow>
`;

exports[`<ModalArrow /> snapshot should match snapshot with className 1`] = `
<ModalArrow
  className="example-class"
  placement="bottom"
>
  <svg
    aria-hidden="true"
    className="md-modal-arrow-svg example-class"
    data-placement="bottom"
    height={12}
    viewBox="0 0 24 12"
    width={24}
    xmlns="http://www.w3.org/svg/2000"
  >
    <defs>
      <clipPath
        id="modal-arrow-cut-stroke-bottom"
      >
        <path
          d="m 0 12 l 10 -10 q 2 -2 4 0 l 10 10"
        />
      </clipPath>
    </defs>
    <path
      clipPath="url(#modal-arrow-cut-stroke-bottom)"
      d="m 0 12 l 10 -10 q 2 -2 4 0 l 10 10"
      data-color="primary"
    />
  </svg>
</ModalArrow>
`;

exports[`<ModalArrow /> snapshot should match snapshot with color 1`] = `
<ModalArrow
  color="quaternary"
  placement="bottom"
>
  <svg
    aria-hidden="true"
    className="md-modal-arrow-svg"
    data-placement="bottom"
    height={12}
    viewBox="0 0 24 12"
    width={24}
    xmlns="http://www.w3.org/svg/2000"
  >
    <defs>
      <clipPath
        id="modal-arrow-cut-stroke-bottom"
      >
        <path
          d="m 0 12 l 10 -10 q 2 -2 4 0 l 10 10"
        />
      </clipPath>
    </defs>
    <path
      clipPath="url(#modal-arrow-cut-stroke-bottom)"
      d="m 0 12 l 10 -10 q 2 -2 4 0 l 10 10"
      data-color="quaternary"
    />
  </svg>
</ModalArrow>
`;

exports[`<ModalArrow /> snapshot should match snapshot with id 1`] = `
<ModalArrow
  id="example-id"
  placement="bottom"
>
  <svg
    aria-hidden="true"
    className="md-modal-arrow-svg"
    data-placement="bottom"
    height={12}
    id="example-id"
    viewBox="0 0 24 12"
    width={24}
    xmlns="http://www.w3.org/svg/2000"
  >
    <defs>
      <clipPath
        id="modal-arrow-cut-stroke-bottom"
      >
        <path
          d="m 0 12 l 10 -10 q 2 -2 4 0 l 10 10"
        />
      </clipPath>
    </defs>
    <path
      clipPath="url(#modal-arrow-cut-stroke-bottom)"
      d="m 0 12 l 10 -10 q 2 -2 4 0 l 10 10"
      data-color="primary"
    />
  </svg>
</ModalArrow>
`;

exports[`<ModalArrow /> snapshot should match snapshot with style 1`] = `
<ModalArrow
  placement="bottom"
  style={
    {
      "color": "pink",
    }
  }
>
  <svg
    aria-hidden="true"
    className="md-modal-arrow-svg"
    data-placement="bottom"
    height={12}
    style={
      {
        "color": "pink",
      }
    }
    viewBox="0 0 24 12"
    width={24}
    xmlns="http://www.w3.org/svg/2000"
  >
    <defs>
      <clipPath
        id="modal-arrow-cut-stroke-bottom"
      >
        <path
          d="m 0 12 l 10 -10 q 2 -2 4 0 l 10 10"
        />
      </clipPath>
    </defs>
    <path
      clipPath="url(#modal-arrow-cut-stroke-bottom)"
      d="m 0 12 l 10 -10 q 2 -2 4 0 l 10 10"
      data-color="primary"
    />
  </svg>
</ModalArrow>
`;
